File: Leetcode/0904 Fruit Into Baskets.java
import java.util.*;

class Solution {
    private static final int BASKET_SIZE = 2;

    public int totalFruit(int[] fruits) {
        int left = 0, maxWindowSize = 0;
        Map<Integer, Integer> lastSeenIndex = new HashMap<>();

        for (int right = 0; right < fruits.length; ++right) {
            int fruit = fruits[right];

            if (!lastSeenIndex.containsKey(fruit) && lastSeenIndex.size() == BASKET_SIZE) {
                int minIndex = Integer.MAX_VALUE, minIndexFruit = 0;
                for (var fruitIndex : lastSeenIndex.entrySet()) {
                    int index = fruitIndex.getValue();
                    if (index < minIndex) {
                        minIndex = index;
                        minIndexFruit = fruitIndex.getKey();
                    }
                }
                left = minIndex + 1;
                lastSeenIndex.remove(minIndexFruit);
            }
            
            lastSeenIndex.put(fruit, right);
            int windowSize = right - left + 1;
            maxWindowSize = Math.max(maxWindowSize, windowSize);
        }
        return maxWindowSize;
    }
}

File: Leetcode/0904 Fruit Into Baskets.py
class Solution:
    BASKET_SIZE = 2

    def totalFruit(self, fruits):
        left, right, max_window_len = 0, 0, 0
        last_index = {}

        for right, fruit in enumerate(fruits):
            if fruit not in last_index and len(last_index) == Solution.BASKET_SIZE:
                min_last_index_fruit = min(last_index, key=last_index.get)
                left = last_index[min_last_index_fruit] + 1
                last_index.pop(min_last_index_fruit)

            last_index[fruit] = right
            window_len = right - left + 1
            max_window_len = max(max_window_len, window_len)

        return max_window_len
